agent/consul/multilimiter/multilimiter.go

```diff
@@ -267,8 +234,27 @@ func (m *MultiLimiter) reconcileLimitedOnce(now time.Time, reconcileCheckLimit t
 		}
 		k, v, ok = iter.Next()
 	}
-	limiters = txn.Commit()
-	m.limiters.CompareAndSwap(storedLimiters, limiters)
+}
+
+func (m *MultiLimiter) cleanLimiters(now time.Time, reconcileCheckLimit time.Duration, txn *radix.Txn) {
+
+	iter := txn.Root().Iterator()
+	k, v, ok := iter.Next()
+	// remove all expired limiters
+	for ok {
+		if t, ok := v.(*Limiter); ok {
+			if t.limiter != nil {
+				lastAccess := t.lastAccess.Load()
+				lastAccessT := time.UnixMilli(lastAccess)
+				diff := now.Sub(lastAccessT)
+
+				if diff > reconcileCheckLimit {
+					txn.Delete(k)
+				}
+			}
+		}
+		k, v, ok = iter.Next()
+	}
```

Comment:
Personal style (please feel free to ignore):

I tend to find that handling all of the type-cast failures and `nil` checking up-front (and reducing nesting) makes this kind of thing easier to read. 

```suggestion
	// remove all expired limiters
	for k, v, ok := iter.Next(); ok; k, v, ok = iter.Next() {
		t, isLimiter := v.(*Limiter)
		if !isLimiter || t.limiter == nil {
			continue
		}

		lastAccess := time.UnixMilli(t.lastAccess.Load())
		if now.Sub(lastAccess) > reconcileCheckLimit {
			txn.Delete(k)
		}
	}

```

Reply:
Similarly, we could reduce the nesting in `reconcileConfig` too:

```Go
for k, v, ok := iter.Next(); ok; k, v, ok = iter.Next() {
	lim, isLimiter := v.(*Limiter)
	if !isLimiter || lim.limiter == nil {
		continue
	}

	// Grab the config for the limiter's prefix.
	prefix, _ := splitKey(k)
	val, exists := m.limitersConfigs.Load().Get(prefix)
	if !exists {
		continue
	}
	config, isConfig := val.(*LimiterConfig)
	if !isConfig || config == nil {
		continue
	}

	// Apply it to the limiter.
	if config.isApplied(lim.limiter) {
		continue
	}
	limiter := Limiter{limiter: rate.NewLimiter(config.Rate, config.Burst)}
	limiter.lastAccess.Store(lim.lastAccess.Load())
	txn.Insert(k, &limiter)
}
```